design/baremetal-operator/bmc-events.md

```diff
@@ -0,0 +1,101 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Event Subscription API
+
+Users of bare metal hardware may want to receive events from the
+baseboard management controller (BMC) in order to act on them, such as
+failed hardware, increases in temperature, and other reasons.
+
+## Status
+
+provisional
+
+## Summary
+
+The Redfish standard includes the ability to subscribe to events, which
+will cause hardware events to be sent in a particular format to a target
+URI. This design document describes a Metal3 API for configuring a
+subscription. While Redfish is the primary target for this design, the
+Ironic API is vendor-neutral and seeks to provide a unified interface
+for configuring events.
+
+## Motivation
+
+Some environments run workloads that need to deal with potential
+failures or environmental changes quicker than they would get an alert
+from other kind of eventing software. For example, some workloads may
+have a sidecar container that knows how to deal with an alert that a
+particular network interface went down, or that the CPU temperature
+reached a certain threshold.
+
+### Goals
+
+- Provide an API to manage subscriptions to events
+
+### Non-Goals
+
+- Any kind of event polling
+- Software for processing the events, i.e. any webhook
+- BMC's beyond Redfish for now
+
+## Proposal
+
+### User Stories
+
+- As a user, I'd like to configure a my BMC to send events to a target URL.
+- As a user, I'd like to filter the types of events the BMC sends to my target URL.
+- As a user, I'd like to provide context to a particular event subscription.
+- As a user, I'd like to provide arbitrary headers.
+
+## Design Details
+
+### Implementation Details
+
+![Sketch of BMCEventSubscription API](./bmc-events.png)
+
+- A BMCEventSubscription resource represents a subscription to the events generated
+  by a specific BMC.
+- Ironic will manage configuring the subscription, using a new API for managing them.
+- The BMCEventSubscription with maintain a reference to a BareMetalHost.
+- The BMCEventSubscription must have a way to inject headers, to allow for tokens,
+  basic auth, etc.
+
+#### Open Questions
```

Comment:
We need to decide when to register the notifications, the Ironic design allows them at any time, but I think we probably want to only configure them when a host is `active`? And remove them when it's not?

Reply:
I don't think we should tie the event subscription to the provisioning state. The goal downstream is to support monitoring hardware events. It doesn't matter whether the hosts being monitored are in use yet or not.

Does relying on Ironic to proxy the subscription calls mean that we need to wait for a host to be successfully registered? So the ironic node is in the "manageable" state? Can anything go wrong in that process that would affect provisioning but not affect the event subscription? Or does Ironic basically check the credentials and that's it?

Reply:
Subscription it's an entity that could have its own state (active/not active), and I also agree should be independent by the current host provisioning state.

Reply:
If we know we have to not be in the enroll state, we could handle that cleanly with a retry w/o error. Other states should keep retrying and report an error if one is produced. Maybe we can reuse the backoff logic that @andfasano added so we don't thrash ironic with requests.